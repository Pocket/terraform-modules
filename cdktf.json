{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
