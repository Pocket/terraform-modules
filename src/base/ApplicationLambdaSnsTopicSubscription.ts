import { DataAwsIamPolicyDocument } from '@cdktf/provider-aws/lib/data-aws-iam-policy-document';
import { DataAwsLambdaFunction } from '@cdktf/provider-aws/lib/data-aws-lambda-function';
import { LambdaFunction } from '@cdktf/provider-aws/lib/lambda-function';
import { LambdaPermission } from '@cdktf/provider-aws/lib/lambda-permission';
import {
  SnsTopicSubscription,
  SnsTopicSubscriptionConfig,
} from '@cdktf/provider-aws/lib/sns-topic-subscription';
import { SqsQueue } from '@cdktf/provider-aws/lib/sqs-queue';
import { SqsQueuePolicy } from '@cdktf/provider-aws/lib/sqs-queue-policy';
import { TerraformMetaArguments, TerraformResource } from 'cdktf';
import { Construct } from 'constructs';

/** The config props type of [[`ApplicationLambdaSnsTopicSubscription]] */
export interface ApplicationLambdaSnsTopicSubscriptionProps
  extends TerraformMetaArguments {
  /** The prefix used to help identify related resources */
  name: string;
  /** The SNS topic to subscribe the Lambda to */
  snsTopicArn: string;
  /** The Lambda that should be invoked by incoming messages to the SNS topic */
  lambda: DataAwsLambdaFunction | LambdaFunction;
  /** Tags to apply to the resource(s), where applicable (in this case only the DLQ for the SNS) */
  tags?: { [key: string]: string };
  /** Optional list of resource dependencies */
  dependsOn?: TerraformResource[];
}

/**
 * Creates an SNS to Lambda subscription with a DLQ for any messages
 * that failed to send to the Lambda (e.g. due to permissions error).
 * Automatically adds policies/permissions for the SNS topic to send
 * messages to the DLQ and invoke the Lambda function.
 *
 * Artifacts:
 *  * {@link https://www.terraform.io/docs/providers/aws/r/sns_topic_subscription aws_sns_topic_subscription} Resource to subscribe Lambda to SNS topic
 *  * {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue aws_sqs_queue} Resource (DLQ for the SNS topic)
 *  * {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue_policy aws_sqs_queue_policy} Resource policy for SNS to send messages to DLQ
 *  * {@link https://www.terraform.io/docs/providers/aws/r/lambda_permission aws_lambda_permission} Resource permission for SNS to invoke Lambda
 */
export class ApplicationLambdaSnsTopicSubscription extends Construct {
  /** the {@link https://www.terraform.io/docs/providers/aws/r/sns_topic_subscription aws_sns_topic_subscription} resource */
  public readonly snsTopicSubscription: SnsTopicSubscription;
  /** the {@link https://www.terraform.io/docs/providers/aws/r/sqs_queue aws_sqs_queue} (DLQ) resource */
  public readonly snsTopicDlq: SqsQueue;

  constructor(
    scope: Construct,
    private name: string,
    private config: ApplicationLambdaSnsTopicSubscriptionProps
  ) {
    super(scope, name);

    this.snsTopicDlq = this.createSqsSubscriptionDlq();
    this.snsTopicSubscription = this.createSnsTopicSubscription(
      this.snsTopicDlq
    );
    this.createDlqPolicy(this.snsTopicDlq);
    this.createLambdaPolicy();
  }

  /**
   * Create a dead-letter queue for failed SNS messages
   * @private
   */
  private createSqsSubscriptionDlq(): SqsQueue {
    return new SqsQueue(this, 'sns-topic-dlq', {
      name: `${this.config.name}-SNS-Topic-DLQ`,
      tags: this.config.tags,
      provider: this.config.provider,
    });
  }

  /**
   * Create an SNS subscription for Lambda
   * @param snsTopicDlq the DLQ for messages that failed to be processed
   * @private
   */
  private createSnsTopicSubscription(
    snsTopicDlq: SqsQueue
  ): SnsTopicSubscription {
    return new SnsTopicSubscription(this, 'sns-subscription', {
      topicArn: this.config.snsTopicArn,
      protocol: 'lambda',
      endpoint: this.config.lambda.arn,
      redrivePolicy: JSON.stringify({
        deadLetterTargetArn: snsTopicDlq.arn,
      }),
      dependsOn: [
        snsTopicDlq,
        this.config.lambda.arn,
        ...(this.config.dependsOn ? this.config.dependsOn : []),
      ],
      provider: this.config.provider,
      tags: this.config.tags,
    } as SnsTopicSubscriptionConfig);
  }

  /**
   * Grant permissions for SNS topic to invoke lambda
   * Cannot be applied to an alias; must use the base lambda function
   */
  private createLambdaPolicy(): void {
    new LambdaPermission(this, `${this.name}-lambda-permission`, {
      principal: 'amazonaws.com',
      action: 'lambda:InvokeFunction',
      functionName: this.config.lambda.functionName,
      sourceArn: this.config.snsTopicArn,
      provider: this.config.provider,
    });
  }

  /**
   * Create IAM policies to allow SNS write to the dead-letter queue
   * @param snsTopicDlq the SQS resource (used as DLQ) to grant permissions on
   * @private
   */
  private createDlqPolicy(snsTopicDlq: SqsQueue): void {
    const queue = { name: 'sns-dlq', resource: snsTopicDlq };
    const policy = new DataAwsIamPolicyDocument(
      this,
      `${queue.name}-policy-document`,
      {
        statement: [
          {
            effect: 'Allow',
            actions: ['sqs:SendMessage'],
            resources: [queue.resource.arn],
            principals: [
              {
                identifiers: ['amazonaws.com'],
                type: 'Service',
              },
            ],
            condition: [
              {
                test: 'ArnEquals',
                variable: 'aws:SourceArn',
                values: [this.config.snsTopicArn],
              },
            ],
          },
        ],
        dependsOn: [queue.resource] as TerraformResource[],
        provider: this.config.provider,
      }
    ).json;

    new SqsQueuePolicy(this, `${queue.name}-policy`, {
      queueUrl: queue.resource.url,
      policy: policy,
      provider: this.config.provider,
    });
  }
}
