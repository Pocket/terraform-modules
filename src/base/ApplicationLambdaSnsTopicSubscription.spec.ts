import { Testing } from 'cdktf';
import { lambdafunction } from '@cdktf/provider-aws';
import { ApplicationLambdaSnsTopicSubscription } from './ApplicationLambdaSnsTopicSubscription';

describe('ApplicationSqsSnsTopicSubscription', () => {
  const getConfig = (stack) => ({
    name: 'test-sns-subscription',
    snsTopicArn: 'arn:aws:sns:TopicName',
    lambdaArn: new lambdafunction.DataAwsLambdaFunction(stack, 'lambda', {
      functionName: 'test-lambda',
    }).arn,
  });

  it('renders an Lambda <> SNS subscription without tags', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationLambdaSnsTopicSubscription(
        stack,
        'lambda-sns-subscription',
        getConfig(stack)
      );
    });
    expect(synthed).toMatchSnapshot();
  });

  it('renders an SQS SNS subscription with tags', () => {
    const synthed = Testing.synthScope((stack) => {
      new ApplicationLambdaSnsTopicSubscription(
        stack,
        'lambda-sns-subscription',
        {
          ...getConfig(stack),
          tags: { hello: 'there' },
        }
      );
    });
    expect(synthed).toMatchSnapshot();
  });
});
