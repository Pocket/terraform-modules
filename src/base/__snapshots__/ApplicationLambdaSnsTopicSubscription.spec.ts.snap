// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationSqsSnsTopicSubscription renders an Lambda <> SNS subscription without tags 1`] = `
"{
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"lambda-sns-subscription_sns-dlq-policy-document_8DAB362F\\": {
        \\"depends_on\\": [
          \\"aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199\\"
        ],
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sqs:SendMessage\\"
            ],
            \\"condition\\": [
              {
                \\"test\\": \\"ArnEquals\\",
                \\"values\\": [
                  \\"arn:aws:sns:TopicName\\"
                ],
                \\"variable\\": \\"aws:SourceArn\\"
              }
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"sns.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ],
            \\"resources\\": [
              \\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.arn}\\"
            ]
          }
        ]
      }
    },
    \\"aws_lambda_function\\": {
      \\"sqs\\": {
        \\"function_name\\": \\"test-lambda\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_sns_topic_subscription\\": {
      \\"lambda-sns-subscription_1ED18AE9\\": {
        \\"depends_on\\": [
          \\"aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199\\",
          \\"data.aws_lambda_function.sqs\\"
        ],
        \\"endpoint\\": \\"\${data.aws_lambda_function.sqs.arn}\\",
        \\"protocol\\": \\"lambda\\",
        \\"redrive_policy\\": \\"{\\\\\\"deadLetterTargetArn\\\\\\":\\\\\\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.arn}\\\\\\"}\\",
        \\"topic_arn\\": \\"arn:aws:sns:TopicName\\"
      }
    },
    \\"aws_sqs_queue\\": {
      \\"lambda-sns-subscription_sns-topic-dlq_C5D5F199\\": {
        \\"name\\": \\"test-sns-subscription-SNS-Topic-DLQ\\"
      }
    },
    \\"aws_sqs_queue_policy\\": {
      \\"lambda-sns-subscription_sns-dlq-policy_31243636\\": {
        \\"policy\\": \\"\${data.aws_iam_policy_document.lambda-sns-subscription_sns-dlq-policy-document_8DAB362F.json}\\",
        \\"queue_url\\": \\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.url}\\"
      }
    }
  }
}"
`;

exports[`ApplicationSqsSnsTopicSubscription renders an SQS SNS subscription with tags 1`] = `
"{
  \\"data\\": {
    \\"aws_iam_policy_document\\": {
      \\"lambda-sns-subscription_sns-dlq-policy-document_8DAB362F\\": {
        \\"depends_on\\": [
          \\"aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199\\"
        ],
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sqs:SendMessage\\"
            ],
            \\"condition\\": [
              {
                \\"test\\": \\"ArnEquals\\",
                \\"values\\": [
                  \\"arn:aws:sns:TopicName\\"
                ],
                \\"variable\\": \\"aws:SourceArn\\"
              }
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"sns.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ],
            \\"resources\\": [
              \\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.arn}\\"
            ]
          }
        ]
      }
    },
    \\"aws_lambda_function\\": {
      \\"sqs\\": {
        \\"function_name\\": \\"test-lambda\\"
      }
    }
  },
  \\"resource\\": {
    \\"aws_sns_topic_subscription\\": {
      \\"lambda-sns-subscription_1ED18AE9\\": {
        \\"depends_on\\": [
          \\"aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199\\",
          \\"data.aws_lambda_function.sqs\\"
        ],
        \\"endpoint\\": \\"\${data.aws_lambda_function.sqs.arn}\\",
        \\"protocol\\": \\"lambda\\",
        \\"redrive_policy\\": \\"{\\\\\\"deadLetterTargetArn\\\\\\":\\\\\\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.arn}\\\\\\"}\\",
        \\"topic_arn\\": \\"arn:aws:sns:TopicName\\"
      }
    },
    \\"aws_sqs_queue\\": {
      \\"lambda-sns-subscription_sns-topic-dlq_C5D5F199\\": {
        \\"name\\": \\"test-sns-subscription-SNS-Topic-DLQ\\",
        \\"tags\\": {
          \\"hello\\": \\"there\\"
        }
      }
    },
    \\"aws_sqs_queue_policy\\": {
      \\"lambda-sns-subscription_sns-dlq-policy_31243636\\": {
        \\"policy\\": \\"\${data.aws_iam_policy_document.lambda-sns-subscription_sns-dlq-policy-document_8DAB362F.json}\\",
        \\"queue_url\\": \\"\${aws_sqs_queue.lambda-sns-subscription_sns-topic-dlq_C5D5F199.url}\\"
      }
    }
  }
}"
`;
